src/autotester/server/utils/tests/test_setup_files.py

```diff
@@ -0,0 +1,386 @@
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        files_dir = os.path.join(tmp_dir, FILES_DIRNAME)
+        os.mkdir(files_dir)
+        tempfile.NamedTemporaryFile(dir=files_dir)
+        with patch(
+            "autotester.server.utils.redis_management.test_script_directory",
+            return_value=tmp_dir,
+        ):
+            yield tmp_dir
+
+
+@pytest.fixture
+def tmp_script_dir_with_one_dir():
+    """
+    Mock the test_script_directory method and yield a temporary directory which has only one subdirectory
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        files_dir = os.path.join(tmp_dir, FILES_DIRNAME)
+        os.mkdir(files_dir)
+        tempfile.TemporaryDirectory(dir=files_dir)
+        with patch(
+            "autotester.server.utils.redis_management.test_script_directory",
+            return_value=tmp_dir,
+        ):
+            yield tmp_dir
+
+
+@pytest.fixture
+def nested_tmp_script_dir():
+    """
+    Mock the test_script_directory method and yield a temporary directory which has nested file structure
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        files_dir = os.path.join(tmp_dir, FILES_DIRNAME)
+        os.mkdir(files_dir)
+        with tempfile.TemporaryDirectory(dir=files_dir) as sub_dir1:
+            with tempfile.TemporaryDirectory(dir=sub_dir1) as sub_dir2:
+                tempfile.NamedTemporaryFile(dir=sub_dir2)
+                with patch(
+                    "autotester.server.utils.redis_management.test_script_directory",
+                    return_value=tmp_dir,
+                ):
+                    yield tmp_dir
+
+
+@pytest.fixture
+def tmp_script_dir_has_multiple_fd():
+    """
+    Mock the test_script_directory method and yield a temporary directory
+    which has more than one file or directory
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        files_dir = os.path.join(tmp_dir, FILES_DIRNAME)
+        os.mkdir(files_dir)
+        tempfile.TemporaryDirectory(dir=files_dir)
+        tempfile.NamedTemporaryFile(dir=files_dir)
+        with tempfile.TemporaryDirectory(dir=files_dir) as root_dir:
+            with tempfile.TemporaryDirectory(dir=root_dir) as sub_dir:
+                tempfile.NamedTemporaryFile(dir=sub_dir)
+                with patch(
+                    "autotester.server.utils.redis_management.test_script_directory",
+                    return_value=tmp_dir,
+                ):
+                    yield tmp_dir
+
+
+@pytest.fixture
+def t_path():
+    """
+    Returns a temporary directory with one sub directory
+    """
+    with tempfile.TemporaryDirectory() as tests_path:
+        with tempfile.NamedTemporaryFile(dir=tests_path):
+            yield tests_path
+
+
+@pytest.fixture
+def f_path_empty():
+    """
+    Returns an empty temporary directory
+    """
+    files_path = tempfile.mkdtemp()
```

Comment:
here and below ... please don't use mkdtemp and mkstemp unless it's absolutely necessary. These will not be cleaned up properly.

Reply:
The files_path contents are moved into tests_path, this will copy the contents first and delete the source directory. So when I used tempfile.TemporaryDirectory it was already deleted while moving the contents from files_path, when it's looking to delete the temporary directory in the fixture, it raised an error that says "File or Folder does not exist". So I am using mkdtemp and mkstemp.

Reply:
I think that is probably because you were returning from the fixture instead of yielding. If you yield then any code your test runs is still technically in the context of the `with` block. This is what you are doing in most of your other fixtures. 

Also, you're right that if the code runs correctly, the source directory will be removed, but if there is a bug (and the test fails) then the source directory may not be cleaned up. It's not good to rely on the code that you are testing to clean up after your tests.

Reply:
Also, you're right that in some cases, if the code works properly, tempfile will be trying to remove a file or directory that has already been removed by the test code. However, your tests should make sure that any of these temporary files are cleaned up regardless of whether the tests pass or not. This can be done without using mkdtemp and mkstemp.